### ISSUE TITLE
Are keys based on Ed25519 supported in PHPSecLib v3.0.12?

### ISSUE THREAD

**Issue by raghuveer** [phpseclib/phpseclib#1737]
Hi,

While I started implementing  RSA 2048 bit keys in old format for key based authentication through phpseclib, I wanted to know if we can use Ed25519 algorithm through phpseclib, since I read in few places that Ed25519 based keys are generated as per new  OpenSSH private key format.

https://security.stackexchange.com/a/144044
https://stribika.github.io/2015/01/04/secure-secure-shell.html

//key generation in OLD OpenSSH private format for RSA
ssh-keygen -f ~/.ssh/my_key_rsa -t rsa -b 4096 -o -a 100 -q -N "chosen_passphrase" -l "Fingerprint"

//key generation in NEW OpenSSH private format for ed25519
ssh-keygen -f ~/.ssh/my_key_ed25519 -t ed25519 -o -a 100 -q -N "chosen_passphrase" -l "Fingerprint"

1) since I read in one of the github issues (https://github.com/phpseclib/phpseclib/issues/1686#issuecomment-899528494) of phpseclib that phpseclib supports elliptic curves, please share if we can use ed25519 keys that are generated using above ssh-keygen command.

2) Also tell me, if "rounds of key derivations" is accepted to be 100 or we need to reduce that number when generating either RSA or ED25519 keys for phpseclib usage?

3) One other query is that, if we can use RSA 4096 bit keys for older systems like RHEL 6/CentOS 6 and RHEL 7/CentOS 7 kind of?

please share your inputs

Thank you

**Comment by terrafrost** [comment]
The commands that you posted don't work for me. idk what some of those options even mean. In consulting https://linux.die.net/man/1/ssh-keygen ,

- **-o** does not appear to be discussed at all
- **-a** _trials_: Specifies the number of primality tests to perform when screening DH-GEX candidates using the **-T** command.
  Not relevant to either rsa or ed25519 keys
- **-l'** Show fingerprint of specified public key file. Private RSA1 keys are also supported. For RSA and DSA keys ssh-keygen tries to find the matching public key file and prints its fingerprint. If combined with -v, an ASCII art representation of the key is supplied with the fingerprint.
  That shows the fingerprint of the current key on the file system and using it makes it so that a new key isn't written. It's analogous to `ssh-keygen -f /path/to/test.pem -l`
- `"Fingerprint"`.
  idk what this is supposed to be. I get a `Too many arguments.` error unless this line is removed.

Here's what I used:

```
ssh-keygen -f test.pem -t ed25519 -a 100 -q -N "chosen_passphrase"
```

> 1. since I read in one of the github issues (PublicKeyLoader leaking memory on SFTP login #1686 (comment)) of phpseclib that phpseclib supports elliptic curves, please share if we can use ed25519 keys that are generated using above ssh-keygen command.

phpseclib does support ed25519 keys and it supports OpenSSH private keys but phpseclib does not support _encrypted_ OpenSSH private keys. An explanation as to why can be found in the comments of Crypt/Common/Formats/Keys/OpenSSH.php:

https://github.com/phpseclib/phpseclib/blob/13b5ad9593072f1b2eca9047af722025e4590849/phpseclib/Crypt/Common/Formats/Keys/OpenSSH.php#L95-L120

> Also tell me, if "rounds of key derivations" is accepted to be 100 or we need to reduce that number when generating either RSA or ED25519 keys for phpseclib usage?

As elaborated above, "rounds of key derivations" is not applicable to ed25519 or anything except for DH-GEX. ed25519 doesn't involve any user-generated primes anyway. The private key is 32 random bytes. Any combination of random bytes. This is in contrast to RSA wherein _p_ and _q_ (which are used to generate _n_) are prime numbers. Not every combination of bits is going to be a prime number. But from the comments I gather not even ssh-keygen's RSA key generation uses "rounds of key derivations".

> One other query is that, if we can use RSA 4096 bit keys for older systems like RHEL 6/CentOS 6 and RHEL 7/CentOS 7 kind of?

Yes, RSA 4096 keys are supported by phpseclib

**Comment by terrafrost** [comment]
The stackoverflow link you provided discusses `-o` and `-a` in more detail:

> The -a 100 option specifies 100 rounds of key derivations, making your key's password harder to brute-force.

No, keys with this parameter option are not supported. I wish they could be but, like I said in my last post, encrypted OpenSSH keys are not supported, irrespective of how many rounds there are of key derivations.

**Comment by raghuveer** [comment]
@terrafrost Thanks for your detailed explanation, I took reference of info in the following page when trying to create the command that is mentioned above, https://www.ssh.com/academy/ssh/keygen

Even if I remove the key derivations concept, can I generate ED25519 keys without encrypted OpenSSH key format?

Tried the following,

ssh-keygen -f test.pem -t ed25519 -q -N "chosen_passphrase"

The command created the private key with following header,

-----BEGIN OPENSSH PRIVATE KEY-----

P.S. While I will try to use RSA 4096 bit keys on old servers, can you please help with command to generate ED25519 keys without encrypted OpenSSH keys format?

Thank you

**Comment by terrafrost** [comment]
> Even if I remove the key derivations concept, can I generate ED25519 keys without encrypted OpenSSH key format?
>
>Tried the following,
>
>ssh-keygen -f test.pem -t ed25519 -q -N "chosen_passphrase"

Yah - unencrypted keys are supported. To do an unencrypted key remove the `-N "chosen_passphrase"` bit and when it gives you a `Enter passphrase (empty for no passphrase):` prompt then hit Enter.

You could also use PKCS8 by doing `-m pkcs8` (or `m pem`) and then set a password (`-N "chosen_passphrase"`) with a few caveats. This seems to work fine for RSA keys. eg.

```
ssh-keygen -f test.pem -t rsa -b 2048 -q -m pkcs8 # no password
ssh-keygen -f test3.pem -t rsa -b 2048 -q -m pkcs8 -N password # with password
```
This, however, does not appear to work with ed25519. At least not on OpenSSH_8.2p1 Ubuntu-4ubuntu0.2, OpenSSL 1.1.1f  31 Mar 2020. Most likely because they're using the older [RFC5208 Public-Key Cryptography Standards (PKCS) #8](https://www.rfc-editor.org/rfc/rfc5208) instead of the newer [RFC5958 Asymmetric Key Packages](https://www.rfc-editor.org/rfc/rfc5958). The latter deprecates the former (RFC5958 says "_Obsoletes: 5208_" and RFC5208 says "_Obsoleted by: 5958_") but maybe OpenSSH never got the memo. phpseclib implements the RFC5958 - not RFC5208.

ie. `ssh-keygen -f test3.pem -t ed25519 -q -m pkcs8` still generates an OpenSSH formatted key, even in-spite of the `-m pkcs8`. So I guess where ssh-keygen is concerned, just don't use encrypted Ed25519 keys or something...

**Comment by terrafrost** [comment]
From https://www.openssh.com/txt/release-6.5:

```
 * Add a new private key format that uses a bcrypt KDF to better
   protect keys at rest. This format is used unconditionally for
   Ed25519 keys, but may be requested when generating or saving
   existing keys of other types via the -o ssh-keygen(1) option.
   We intend to make the new format the default in the near future.
   Details of the new format are in the PROTOCOL.key file.
```
No newer changelog entries discuss any change to this behavior so it looks like if you want to do Ed25519 with OpenSSH you have to use their format. Which means no password protection if you want to use phpseclib 😐

**Comment by raghuveer** [comment]
I will try creating keys as per RSA and ED25519 algorithms based on commands you suggested above and see how that works with phpseclib.

Can we expect in future passphrase support for ED25519?

I will post update after testing the above, thank you for your inputs @terrafrost

**Comment by terrafrost** [comment]
> I will try creating keys as per RSA and ED25519 algorithms based on commands you suggested above and see how that works with phpseclib.
>
> Can we expect in future passphrase support for ED25519?

Probably never.

If you're willing to wait 30 seconds to 1 minute for a key to load, sure, I can do it - I can add passphrase support for encrypted OpenSSH keys (which is more broadly what you're asking for). But I doubt that taking 30s to 1m to load a private key is acceptable for you or anyone.

I mean, I can revisit it. Maybe there's something I missed last time. Maybe a fresh perspective would be beneficial. But that's a [hail mary](https://en.wikipedia.org/wiki/Hail_Mary_pass).

I included a code snippet discussing some of the technical challenges in an earlier post in this thread but here it is again:

https://github.com/phpseclib/phpseclib/blob/13b5ad9593072f1b2eca9047af722025e4590849/phpseclib/Crypt/Common/Formats/Keys/OpenSSH.php#L95-L120

**Comment by raghuveer** [comment]
@terrafrost , tried RSA 4096 bit key, based on syntax shared by you for RSA key with passphrase scenario,

`ssh-keygen -f /home/username/.ssh/test3rsa4096withpass.pem -t rsa -b 4096 -q -m pkcs8 -N password # with password`

This worked.

please tell me, if we can handle ED25519 with passphrase in the scope of OpenSSH encrypted keys, if we use libsodium etc dependencies?

my apologies for asking this question in different scopes, the reasons are while RSA 4096 bit key with passphrase works for now, we might need to shift to ED25519 with passphrase in next few years and that time, we will be fine to install a required library on the server over thinking about alternatives with phpseclib offering good set of features, strictly in compliance and security perspective.

Please check and tell if we can leverage external dependencies, so those who require them will install those dependencies and rest of them use RSA keys with passphrase and kind of...

Thank you.

**Comment by terrafrost** [comment]
libsodium won't help with encrypted OpenSSH keys. I elaborate here:

https://github.com/phpseclib/phpseclib/blob/13b5ad9593072f1b2eca9047af722025e4590849/phpseclib/Crypt/Common/Formats/Keys/OpenSSH.php#L95-L120

TLDR OpenSSH customized bcrypt in a non-standardized way so you either need to do this in pure-PHP or you don't do it. Well I mean I guess you could use ssh-keygen to remove the PW via `exec()` calls but short of that there's not a ton that can be done

**Comment by raghuveer** [comment]
ED25519 algorithm worked without passphrase, as you posted earlier,

key generation:
ssh-keygen -f /home/cccv3/.ssh/test3ed25519withoutpass.pem -t ed25519 -q -m pkcs8

in phpseclib:
$key = \phpseclib3\Crypt\PublicKeyLoader::load(file_get_contents('/home/cccv3/.ssh/test3ed25519withoutpass.pem'), '');

signature started with:
-----BEGIN OPENSSH PRIVATE KEY-----

**Comment by terrafrost** [comment]
OpenSSH encrypted private keys are now supported as of this commit:

https://github.com/phpseclib/phpseclib/commit/0c8994bbadffd45cd979a1354258b14ec68a025d
